modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -259,38 +265,136 @@ protected void doStop() {
     @Override
     protected void doClose() {}
 
-    private static final class SuccessfulRequestTracker implements HttpPipelinePolicy {
-        private static final Logger logger = LogManager.getLogger(SuccessfulRequestTracker.class);
+    static class RequestMetrics {
+        private volatile long timeToResponseInMillis = -1;
+        private volatile int requestCount;
+        private volatile int errorCount;
+        private volatile int throttleCount;
+        private volatile int statusCode;
+
+        int getRequestCount() {
+            return requestCount;
+        }
+
+        int getErrorCount() {
+            return errorCount;
+        }
+
+        int getStatusCode() {
+            return statusCode;
+        }
+
+        int getThrottleCount() {
+            return throttleCount;
+        }
+
+        /**
+         * Time to response, or -1 if response was never received
+         */
+        long getTimeToResponseInMillis() {
+            return timeToResponseInMillis;
+        }
+
+        @Override
+        public String toString() {
+            return "RequestMetrics{"
+                + "timeToResponseInMillis="
+                + timeToResponseInMillis
+                + ", requestCount="
+                + requestCount
+                + ", errorCount="
+                + errorCount
+                + ", throttleCount="
+                + throttleCount
+                + ", statusCode="
+                + statusCode
+                + '}';
+        }
+    }
+
+    private enum RetryMetricsTracker implements HttpPipelinePolicy {
+        INSTANCE;
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+            Optional<Object> metricsData = context.getData(RequestMetricsTracker.ES_REQUEST_METRICS_CONTEXT_KEY);
+            if (metricsData.isPresent() == false) {
+                assert false : "No metrics object associated with request " + context.getHttpRequest();
+                return next.process();
+            }
+            RequestMetrics metrics = (RequestMetrics) metricsData.get();
+            metrics.requestCount++;
+            return next.process().doOnError(throwable -> {
+                logger.debug("Detected error in RetryMetricsTracker", throwable);
+                metrics.errorCount++;
+            }).doOnSuccess(response -> {
+                if (isErrorCode(response.getStatusCode())) {
+                    metrics.errorCount++;
+                    if (response.getStatusCode() == RestStatus.TOO_MANY_REQUESTS.getStatus()) {
+                        metrics.throttleCount++;
```

Comment:
I did check that, from what I could see, Azure seems to always throttle with a 429.

See https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/request-limits-and-throttling#error-code, I will add that link in a comment.

There are [retries for more than just throttling exceptions](https://learn.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines), but I think we're counting specifically throttles in this metric.

Reply:
AWS client contains more logic because [they use AWS error codes as well as response status](https://github.com/aws/aws-sdk-java/blob/864baf847e35b6c987ad08d81738050d329c515c/aws-java-sdk-core/src/main/java/com/amazonaws/retry/RetryUtils.java#L104) to indicate throttling by the look of things

Reply:
Added link to docs in 27697f9fcdef570834b6973dda6c04acbedfbc78

Reply:
Thanks a lot for checking and the links